oculus_ros2/src/sonar_viewer.cpp

```diff
@@ -0,0 +1,194 @@
+  int kh = psf.cols;
+  cv::Mat psf_padded = cv::Mat::zeros(img.size(), img.type());
+  psf.copyTo(psf_padded(cv::Rect(0, 0, kh, kw)));
+
+  // compute (padded) psf's DFT
+  cv::Mat psf_f;
+  cv::dft(psf_padded, psf_f, cv::DFT_COMPLEX_OUTPUT, kh);
+
+  cv::Mat psf_f_2;
+  cv::pow(psf_f, 2, psf_f_2);
+  cv::transform(psf_f_2, psf_f_2, cv::Matx12f(1, 1));
+
+  double noise = 0.001;
+  cv::Mat ipsf_f(psf_f.size(), CV_64FC2);
+  for (int i = 0; i < psf_f.rows; i++) {
+    for (int j = 0; j < psf_f.cols; j++) {
+      // compute element-wise division
+      cv::Vec2d val = psf_f.at<cv::Vec2d>(i, j) / (psf_f_2.at<double>(i, j) + noise);
+      // store result in ipsf_f
+      ipsf_f.at<cv::Vec2d>(i, j) = val;
+    }
+  }
+  cv::Mat result_f;
+  cv::mulSpectrums(img_f, ipsf_f, result_f, 0);
+
+  cv::Mat result;
+  cv::idft(result_f, result, cv::DFT_SCALE | cv::DFT_REAL_OUTPUT);
+
+  cv::Mat result_shifted_rows(result.size(), result.type());
+  int shift = ceil(kw / 2.0);
+  // similar to roll with shift = -1
+  for (int i = 0; i < result.rows; i++) result.row(i).copyTo(result_shifted_rows.row((i + shift) % result_shifted_rows.rows));
+
+  shift = ceil(kh / 2.0);
+  for (int i = 0; i < result.cols; i++) result_shifted_rows.col(i).copyTo(result.col((i + shift) % result.cols));
+  result.setTo(0, result < 0);
+  cv::normalize(result, result, 0, 1, cv::NORM_MINMAX);
+  cv::imshow("result", result);
+  cv::waitKey(1);
+
+  /*std::stringstream file_name;
+  file_name << "/home/jaouadros/catkin_ws/src/Sonar_processing_display/tests/results/before/image_" << image_i << ".jpg";
+  cv::normalize(img, img, 0, 255, cv::NORM_MINMAX);
+  cv::imwrite(file_name.str(), img);
+  std::stringstream file_name2;
+  cv::normalize(result, result, 0, 255, cv::NORM_MINMAX);
+  file_name2 << "/home/jaouadros/catkin_ws/src/Sonar_processing_display/tests/results/after/image_" << image_i << ".jpg";
+  cv::imwrite(file_name2.str(), result);
+  image_i++;*/
+}
+
+void gammaCorrection(const cv::Mat& src, cv::Mat& dst, const float gamma) {
+  float invGamma = 1 / gamma;
+
+  cv::Mat table(1, 256, CV_8U);
+  uchar* p = table.ptr();
+  for (int i = 0; i < 256; ++i) {
+    p[i] = (uchar)(pow(i / 255.0, invGamma) * 255);
+  }
+
+  LUT(src, table, dst);
+}
+
+std::string type2str(int type) {
+  std::string r;
+
+  uchar depth = type & CV_MAT_DEPTH_MASK;
+  uchar chans = 1 + (type >> CV_CN_SHIFT);
+
+  switch (depth) {
+    case CV_8U:
+      r = "8U";
+      break;
+    case CV_8S:
+      r = "8S";
+      break;
+    case CV_16U:
+      r = "16U";
+      break;
+    case CV_16S:
+      r = "16S";
+      break;
+    case CV_32S:
+      r = "32S";
+      break;
+    case CV_32F:
+      r = "32F";
+      break;
+    case CV_64F:
+      r = "64F";
+      break;
+    default:
+      r = "User";
+      break;
+  }
+
+  r += "C";
+  r += (chans + '0');
+
+  return r;
+}
+
+int grid_presence_counter[36][25] = {0};
+int grid_absence_counter[36][25] = {0};
+int frames_counter = 0;
```

Comment:
I think this is only used to filter the image. But the image is never filtered at the moment, so it is not useful (unless we want to add a filter option in the package). Moreover, I don't think that a gamma correction is a relevant filter from the oculus sonar data since there is already a gamma filter in the sonar software.

Reply:
`gammaCorrection` is to adjust the brightness and contrast of an image. `type2str` is to print the image type. The other variables were used for obstacle detection algorithm.
We don't need all of that.